src/praisonai/praisonai/gateway/exec_approval.py

```
@@ -47,40 +111,411 @@ class Resolution:
+        sig_keys = {"", sig} if sig else {""}
+        placeholders_a = ",".join("?" for _ in agent_keys)
+        placeholders_s = ",".join("?" for _ in sig_keys)
+        with self._lock, closing(self._connect()) as conn:
+            row = conn.execute(
+                f"""
+                SELECT 1 FROM allow_grants
+                WHERE tool_name = ?
+                  AND agent_id IN ({placeholders_a})
+                  AND arg_signature IN ({placeholders_s})
+                LIMIT 1
+                """,
```

Ruff (0.15.20)

[error] Possible SQL injection vector through string-based query construction

(S608)